src/file/reader.rs

```diff
@@ -285,21 +285,45 @@ impl PageReader for SerializedPageReader {
   fn get_next_page(&mut self) -> Result<Option<Page>> {
     while self.seen_num_values < self.total_num_values {
       let page_header = self.read_page_header()?;
-      let compressed_len = page_header.compressed_page_size as usize;
-      let uncompressed_len = page_header.uncompressed_page_size as usize;
-      let mut buffer = vec![0; compressed_len];
+
+      // when processing data page v2, depending on enabled compression for the page, we should
+      // account for uncompressed data ('offset') of repetition and definition levels.
+      let (offset, can_apply_compression) = if page_header.data_page_header_v2.is_some() {
+        let header_v2 = page_header.data_page_header_v2.as_ref().unwrap();
+        let offset = header_v2.definition_levels_byte_length + header_v2.repetition_levels_byte_length;
+        // when is_compressed flag is missing the page is considered compressed
+        let is_compressed = header_v2.is_compressed.unwrap_or(true);
+        (offset as usize, is_compressed)
+      } else {
+        // we always use 0 offset for other pages
+        // `true` flag means that compression will be applied if decompressor is defined
+        (0, true)
+      };
+
+      let compressed_len = page_header.compressed_page_size as usize - offset;
+      let uncompressed_len = page_header.uncompressed_page_size as usize - offset;
+      // we still need to read all bytes from buffered stream
+      let mut buffer = vec![0; offset + compressed_len];
       self.buf.read_exact(&mut buffer)?;
 
       // TODO: page header could be huge because of statistics. We should
       // set a maximum page header size and abort if that is exceeded.
       if let Some(decompressor) = self.decompressor.as_mut() {
-        let mut decompressed_buffer = vec!();
-        let decompressed_size = decompressor.decompress(&buffer, &mut decompressed_buffer)?;
-        if decompressed_size != uncompressed_len {
-          return Err(general_err!("Actual decompressed size doesn't \
-            match the expected one ({} vs {})", decompressed_size, uncompressed_len));
+        if can_apply_compression {
+          let mut decompressed_buffer = vec!();
+          let decompressed_size = decompressor.decompress(&buffer[offset..], &mut decompressed_buffer)?;
+          if decompressed_size != uncompressed_len {
+            return Err(general_err!("Actual decompressed size doesn't \
+              match the expected one ({} vs {})", decompressed_size, uncompressed_len));
+          }
+          if offset == 0 {
+            buffer = decompressed_buffer;
+          } else {
+            // prepend saved offsets to the buffer
+            buffer = buffer[..offset].to_vec();
```

Comment:
nit: replace by `truncate()`?

Reply:
Thanks, will change! I did not know there was such function:)